File: c/mkpass.c
#include <stdio.h>
#include <string.h>
#include <stdbool.h>
#include <stdlib.h>
#include <sys/time.h>

const char VERSION[] = "0.1.0";
const int MAX_TRIES = 10;

const char UPPER[] = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
const char LOWER[] = "abcdefghijklmnopqrstuvwxyz";
const char NUMBER[] = "0123456789";
const char SYMBOL[] = "`~!@#$%^&*()_+-=[]\\{}|;':\",./<>?";

struct options {
  short upper;
  short lower;
  short number;
  short symbol;
  int length;
};

struct options default_options = {
  .upper = true,
  .lower = true,
  .number = true,
  .symbol = true,
  .length = 16
};

struct generator {
  struct options opts;
  char *charset;
};

struct generator new_generator(struct options opts) {
  struct timeval t;
  gettimeofday(&t, NULL);
  srand(t.tv_sec + t.tv_usec);
  struct generator gen;
  gen.opts = opts;
  char *cs = NULL;
  if (opts.upper) {
    int sz = 0;
    if (cs != NULL) {
      sz = strlen(cs);
    }
    void *newp = realloc(cs, sz + strlen(UPPER));
    if (newp == NULL) {
      fprintf(stderr, "Failed to allocate memory for upper\n");
      exit(1);
    }
    cs = newp;
    strcat(cs, UPPER);
  }
  if (opts.lower) {
    int sz = 0;
    if (cs != NULL) {
      sz = strlen(cs);
    }
    void *newp = realloc(cs, sz + strlen(LOWER));
    if (newp == NULL) {
      fprintf(stderr, "Failed to allocate memory for lower\n");
      exit(1);
    }
    cs = newp;
    strcat(cs, LOWER);
  }
  if (opts.number) {
    int sz = 0;
    if (cs != NULL) {
      sz = strlen(cs);
    }
    void *newp = realloc(cs, sz + strlen(NUMBER));
    if (newp == NULL) {
      fprintf(stderr, "Failed to allocate memory for number\n");
      exit(1);
    }
    cs = newp;
    strcat(cs, NUMBER);
  }
  if (opts.symbol) {
    int sz = 0;
    if (cs != NULL) {
      sz = strlen(cs);
    }
    void *newp = realloc(cs, sz + strlen(SYMBOL));
    if (newp == NULL) {
      fprintf(stderr, "Failed to allocate memory for symbol\n");
      exit(1);
    }
    cs = newp;
    strcat(cs, SYMBOL);
  }

  gen.charset = cs;

  return gen;
}

int generate(struct generator gen, char **str) {
  char *np = calloc(sizeof(char), gen.opts.length + 1);
  if (np == NULL) {
    fprintf(stderr, "Failed to allocate memory for generated string\n");
    return 0;
  }

  for (int i = 0; i < gen.opts.length; i++) {
    int x = rand() % strlen(gen.charset);
    np[i] = (char)gen.charset[x];
  }
  *str = np;

  return 1;
}

int main(int argc, char *argv[]) {

  struct generator gen = new_generator(default_options);

  char *pw = NULL;
  if (!generate(gen, &pw)) {
    fprintf(stderr, "Failed to generate string\n");
    return 0;
  }
  printf("%s\n", pw);

  return 1;
}

File: go/mkpass.go
package mkpass

import (
	"crypto/rand"
	"fmt"
	"math/big"
	"strings"
)

const (
	VERSION   string = "0.3.2"
	MAX_TRIES int    = 10

	UPPER  string = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
	LOWER  string = "abcdefghijklmnopqrstuvwxyz"
	NUMBER string = "0123456789"
	SYMBOL string = "`~!@#$%^&*()_+-=[]\\{}|;':\",./<>?"
)

var DefaultOptions = Options{
	Upper:  true,
	Lower:  true,
	Number: true,
	Symbol: true,
	Length: 16,
}

type Options struct {
	Upper  bool
	Lower  bool
	Number bool
	Symbol bool
	Length int
}

type Generator struct {
	opts    Options
	charset string
}

func New(o Options) (Generator, error) {
	g := Generator{
		opts: o,
	}
	nc := 0
	if o.Upper {
		g.charset += UPPER
		nc++
	}
	if o.Lower {
		g.charset += LOWER
		nc++
	}
	if o.Number {
		g.charset += NUMBER
		nc++
	}
	if o.Symbol {
		g.charset += SYMBOL
		nc++
	}
	if o.Length == 0 {
		g.opts.Length = DefaultOptions.Length
	}
	if len(g.charset) == 0 {
		return g, fmt.Errorf("No character classes chosen, characterset is empty")
	}
	if nc > g.opts.Length {
		return g, fmt.Errorf("Length is shorter than number of character classes")
	}
	return g, nil
}

func (g Generator) Generate() (string, error) {
	tries := 0
	var str string

	// loop that breaks when circuit breaker trips or suitable string is generated
	for {
		tries++
		// circuit breaker for uncaught errors that prevent generation of suitable strings
		if tries > MAX_TRIES {
			return "", fmt.Errorf("Failed to generate suitable random string in %d tries", MAX_TRIES)
		}

		str = ""
		// pick random characters from set one at a time
		l := big.NewInt(int64(len(g.charset)))
		for i := 0; i < g.opts.Length; i++ {
			n, err := rand.Int(rand.Reader, l)
			if err != nil {
				return "", fmt.Errorf("Reading random integer: %s", err.Error())
			}
			str += string(g.charset[n.Int64()])
		}

		// check if each used class is included in output
		if g.opts.Upper && !strings.ContainsAny(str, UPPER) {
			continue
		}
		if g.opts.Lower && !strings.ContainsAny(str, LOWER) {
			continue
		}
		if g.opts.Number && !strings.ContainsAny(str, NUMBER) {
			continue
		}
		if g.opts.Symbol && !strings.ContainsAny(str, SYMBOL) {
			continue
		}
		break
	}

	return str, nil
}
